drivers/net/ethernet/intel/idpf/idpf_eth_idc.c

```diff
@@ -0,0 +1,453 @@
+		le16_to_cpu(adapter->caps.max_rx_q);
+	eth_dev->eth_info.caps.q_info.max_txq =
+		le16_to_cpu(adapter->caps.max_tx_q);
+	eth_dev->eth_info.caps.q_info.max_bufq =
+		le16_to_cpu(adapter->caps.max_rx_bufq);
+	eth_dev->eth_info.caps.q_info.max_complq =
+		le16_to_cpu(adapter->caps.max_tx_complq);
+
+	/* Caps */
+	eth_dev->eth_info.caps.csum_caps = adapter->caps.csum_caps;
+	eth_dev->eth_info.caps.seg_caps = adapter->caps.seg_caps;
+	eth_dev->eth_info.caps.hsplit_caps =
+		adapter->caps.hsplit_caps;
+	eth_dev->eth_info.caps.rsc_caps = adapter->caps.rsc_caps;
+	eth_dev->eth_info.caps.rss_caps = adapter->caps.rss_caps;
+	eth_dev->eth_info.caps.other_caps = adapter->caps.other_caps;
+	eth_dev->eth_info.caps.min_sso_packet_len =
+		adapter->caps.min_sso_packet_len;
+	eth_dev->eth_info.caps.max_sg_bufs_per_tx_pkt =
+		adapter->caps.max_sg_bufs_per_tx_pkt;
+}
+
+/**
+ * idpf_eth_idc_device_alloc - allocate auxiliary device
+ * @adapter: Idpf private structure
+ * @vport_type: Vport type
+ *
+ * Returns allocated structure
+ */
+static struct idpf_eth_idc_auxiliary_dev *
+idpf_eth_idc_device_alloc(struct idpf_adapter *adapter,
+			  enum idpf_vport_type vport_type)
+{
+	struct idpf_eth_idc_auxiliary_dev *new_eth_dev;
+	struct idpf_eth_idc_dev_info *dev_info;
+	enum idpf_queue_type queue_type;
+	u16 i, err;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		/* Check pre-allocated one first */
+		if (adapter->adevs[i])
+			continue;
+
+		adapter->adevs[i] =
+			kzalloc(sizeof(struct idpf_eth_idc_auxiliary_dev),
+				GFP_KERNEL);
+		new_eth_dev = adapter->adevs[i];
+		if (!new_eth_dev)
+			goto alloc_exit;
+
+		new_eth_dev->adev.id = ida_alloc(&idpf_eth_idc_ida, GFP_KERNEL);
+		new_eth_dev->eth_info.eth_shared = &adapter->eth_shared;
+		new_eth_dev->eth_info.idx = i;
+
+		/* Set vport type */
+		new_eth_dev->eth_info.vport_type = vport_type;
+
+		if (vport_type == IDPF_DEFAULT_VPORT)
+			queue_type = IDPF_QUEUE_DEFAULT_VPORT;
+		else
+			queue_type = IDPF_QUEUE_DYNAMIC_VPORT;
+
+		dev_info = &new_eth_dev->eth_info;
+		err = idpf_alloc_max_qs(adapter, &dev_info->caps.q_info,
+					queue_type);
+		if (err) {
+			kfree(adapter->adevs[i]);
+			adapter->adevs[i] = NULL;
+			new_eth_dev = NULL;
+		}
+
+		return new_eth_dev;
+	}
+
+alloc_exit:
+	return NULL;
+}
+
+/**
+ * idpf_eth_idc_device_init - Initialize ethernet IDC & device(s)
+ * @adapter: Idpf private structure
+ */
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter)
+{
+	struct device *dev = &adapter->eth_shared.pdev->dev;
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	int err;
+	u16 i;
+
+	if (adapter->adevs) {
+		/* This is reset case so simply invoke add explicitly.
+		 * Direct eth device re-add or probe
+		 */
+		eth_idc_drv = idpf_eth_idc_get_driver();
+		for (i = 0; i < adapter->default_vports; ++i) {
+			eth_dev = adapter->adevs[i];
+			if (!eth_dev)
+				continue;
+
+			eth_idc_drv->adrv.probe(&eth_dev->adev, NULL);
```

Comment:
why is pci driver calling the aux driver probe directly? aux bus should take care of the probe for all its devices all the time. In the reset flow, aux device is present and the probe is already called. Aux driver should receive the reset info through an event and it has to readjust itself one level below its probe.(no difference here like what PCI device/driver does)

Reply:
This will be very difficult to get upstream acceptance i assume, as it changes the very basic flows of bus/driver model. Better we raise the concern to HMA/CP and ask them to prioritize the fix.

Another confusion atleast for me is, you are doing it only in reset flows. Not in driver init flows???So if it is a problem, then it has to exist in both cases?

Reply:
Init flow also follows reset path. The first initialization starts after reset, and mail box init commands etc, so it follows the same path for first time initialization. That's how it has been coded.
I will check if I can remove calling direct probe from PCI driver for successive resets etc. by making some code changes.

Reply:
If hard reset needs to always do probe indirectly (by calling aux dev add), you will have to free the full aux dev structure which houses pointers to the 'eth' adapter and vport_config. 
So we would need to cache vport_config elsewhere (and whatever else in the eth adapter is needed across resets)

Reply:
I have removed the dependency of PCI driver calling probe directly. It now only calls once to initialize/start the device. Later auxiliary device handles the probe by itself. The code in contained in the aux driver. Only the serialization part we have dependency on the PCI driver. Please check the changes, and let me know if I can resolve this.